### ISSUE TITLE
MeanAveragePrecision devices mismatch

### ISSUE THREAD

**Issue by kuntiik** [PyTorchLightning/metrics#943]
Version: TorchMetrics 0.7.3

I have noticed a bug in MeanAveragePrecision() class.
How to reproduce:
The code in pytorch-lightning module is: 
```py
__init__(...):
  self.mAP = MeanAveragePrecision()

validation_step(...):
  self.mAP(preds, targets)  # preds and targets have the same device as pl-module.

validation_epoch_end(...):
  mAP_dict = self.mAP.compute()
  self.log_dict(mAP_dict)
  
```
I get a device mismatch error at line 715 in map.py, when I run my code on a GPU.

I dug deeper and found the problem:
On line 781 is the following code:
```
       # move everything to CPU, as we are faster here
        self.detection_boxes = [box.cpu() for box in self.detection_boxes]
        self.detection_labels = [label.cpu() for label in self.detection_labels]
        self.detection_scores = [score.cpu() for score in self.detection_scores]
        self.groundtruth_boxes = [box.cpu() for box in self.groundtruth_boxes]
        self.groundtruth_labels = [label.cpu() for label in self.groundtruth_labels]
```
This means, that all gt boxes and detections are stored in the CPU, but in the function _evaluate_image(...), when we have no predictions but have GT we call function __evaluate_image_preds_no_gt(...) (line 470 in map.py).
In this function we have:
``` 
       return {
            "dtMatches": torch.zeros((nb_iou_thrs, nb_det), dtype=torch.bool, device=self.device),
            "gtMatches": torch.zeros((nb_iou_thrs, nb_gt), dtype=torch.bool, device=self.device),
            "dtScores": scores_sorted,
            "gtIgnore": gt_ignore,
            "dtIgnore": det_ignore,
        }
```
The problem is the part where we return device=self.device (we moved everything to CPU before).

My questions are:
Have I discovered a bug or do I have an error in my code?
Why has nobody else before encountered this problem, since this bug makes this metric unusable?

Thank you for your answers. 

Edit: I forked the report and edited functions __evaluate_image_gt_no_preds and __evaluate_image_preds_no_gt by replacing device=self.device by device=torch.device('cpu') and so far my code is working.
I think this is a critical bug, which leaves the Class unusable.

**Comment by SkafteNicki** [comment]
Hi @kuntiik,
I have trouble reproducing the error you get. If I run the simple MAP example:
https://github.com/PyTorchLightning/metrics/blob/master/tm_examples/detection_map.py
where I move everything to `cuda` I do not get any errors.

Could you try updating to master, just to make sure we have not solved the issue between 0.7.3 and now?

**Comment by kuntiik** [comment]
Hi,
you tested it on a simple example, which does not catch the error. Try to do the following. Move MeanAveragePrecision to Cuda. Then do update twice: In one of the updates, there has to be a prediction and a ground truth. In the update delete ground truth (pass empty) then try to compute the metrics. I will try to postminimal example asap.

**Comment by kuntiik** [comment]
```{python}
def test_minimal_metrics():
    from torchmetrics.detection import MeanAveragePrecision
    map = MeanAveragePrecision().to(torch.device('cuda:0'))
    preds = [
        dict(
            boxes=torch.Tensor([[258.0, 41.0, 606.0, 285.0]]),
            scores=torch.Tensor([0.536]),
            labels=torch.IntTensor([0]),
        )
    ]
    target = [
        dict(
            boxes=torch.Tensor([[214.0, 41.0, 562.0, 285.0]]),
            labels=torch.IntTensor([0]),
        )
    ]
    map.update(preds, target)
    target = [
        dict(
            boxes=torch.Tensor([]),
            labels=torch.IntTensor([]),
        )
    ]
    map.update(preds, target)
    result = map.compute()
    print(result)

```

**Comment by SkafteNicki** [comment]
Hi @kuntiik,
Thanks for the example that you provided, it does indeed lead to device errors. Going to fix immediately.
